libhif/hif-transaction.c

```diff
@@ -886,6 +889,48 @@ hif_transaction_delete_packages(HifTransaction *transaction,
     return TRUE;
 }
 
+static gchar *
+hif_transaction_get_propagated_reason(HifTransaction *transaction,
+                                      HyGoal goal,
+                                      HifPackage *pkg)
+{
+    HifTransactionPrivate *priv = GET_PRIVATE(transaction);
+
+    /* kernel is always marked as "user" */
+    if (hif_package_is_installonly(pkg))
+        return g_strdup("user");
+
+    /* for updates, propagate updated package's reason from existing yumdb entry */
+    if (hif_package_get_action(pkg) == HIF_STATE_ACTION_DOWNGRADE ||
+        hif_package_get_action(pkg) == HIF_STATE_ACTION_REINSTALL ||
+        hif_package_get_action(pkg) == HIF_STATE_ACTION_UPDATE) {
+        HifPackage *erased_package;
+
+        erased_package = g_hash_table_lookup(priv->erased_by_package_hash, hif_package_get_package_id(pkg));
+        if (erased_package != NULL) {
+            gchar *reason;
+
+            reason = hif_db_get_string(priv->db, erased_package, "reason", NULL);
+            if (reason != NULL) {
```

Comment:
That's hughsie's style to explicitly write out NULL comparisons. Just following existing style here :)